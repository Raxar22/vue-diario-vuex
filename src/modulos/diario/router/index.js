


export default {

    name: 'diario',
    component: () => import(/* webpackChunkName: "diario" */ '@/modulos/diario/layouts/DiarioLayout.vue' ),
    children: [
        {
            path:'',
            name:'NoEntradas',
            component: () => import (/* webpackChunkName: "SinEntradas" */ '@/modulos/diario/views/NingunaEntradaSeleccionada.vue'),
        },
        {
            path:':id',
            name:'Entrada',
            component: () => import (/* webpackChunkName: "Entradas" */ '@/modulos/diario/views/EntradaDiario.vue'),
            props: (route) =>{
                return {
                    id: route.params.id
                }
            }

        }
    ]


}