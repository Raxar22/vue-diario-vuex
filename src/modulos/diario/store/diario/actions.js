
// export const miAccion = async ({commit})=> {

import diarioApi from "@/api/diarioApi"

// }

export const cargarEntradas = async ( {commit} )=> {

    const {data} = await diarioApi.get('/Entradas.json')
    if (!data){
        commit('setEntrada',[])
        return
    }
    const entradas = []
    for (let id of Object.keys(data)){
        entradas.push({
            id,
            ...data[id]
        })
    }
    commit('setEntrada',entradas)
   // console.log (entradas)
}

export const actualizarEntradas = async ( {commit}, entrada )=> {
    //dispara la mutation
    const {fecha,text,imagen} = entrada
    const datoAGuardar =  {fecha,text,imagen}
    await diarioApi.put(`/Entradas/${entrada.id}.json`,datoAGuardar)
    datoAGuardar.id = entrada.id
    commit('updateEntrada',{...datoAGuardar})


}

export const crearEntradas = async ( {commit},entrada )=> {
    const {fecha,text,imagen} = entrada
    const datoAGuardar =  {fecha,text,imagen}
    const entradaNueva = await diarioApi.post(`/Entradas.json`,datoAGuardar)

    datoAGuardar.id = entradaNueva.data.name
    //console.log(entradaNueva.data.name)
    commit('addEntrada',{...datoAGuardar})

    return entradaNueva.data.name

}


export const borrarEntrada = async ( {commit},id )=> {

    await diarioApi.delete(`/Entradas/${id}.json`)
    commit('borrarEntrada',id)
}