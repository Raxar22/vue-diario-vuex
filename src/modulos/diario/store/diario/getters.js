
// export const miGetter = (state)=> {
//  return state
// }

export const obtenerEntradaBySearch = ( state )=> (term = '') => {

    if (term.length===0) return state.entries

    return state.entries.filter (entry => entry.text.toLowerCase().includes(term.toLowerCase()))
}


export const obtenerEntradaById = (state)=> (id='') => {
    const entrada= state.entries.find (entry=> entry.id === id)
    
    if (!entrada) return
    
    return {...entrada}
}