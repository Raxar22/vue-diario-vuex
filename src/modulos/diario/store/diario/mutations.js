
// export const miMutacion =  (state)=> {

// }

export const setEntrada =  (state,entradas)=> {
    state.entries= []
    state.entries= [...state.entries, ...entradas]
    state.isLoading = false
}

export const updateEntrada =  (state,entrada)=> {
    const indice = state.entries.map( e => e.id ).indexOf(entrada.id)
    state.entries[indice] = entrada

}

export const addEntrada =  (state,entrada)=> {
    // console.log(state)
    // console.log(entrada)
    state.entries = [entrada, ...state.entries]
}
export const borrarEntrada =  (state,id)=> {
  /*  const indice = state.entries.map( e => e.id ).indexOf(id)
    state.entries.splice(indice,1)*/


    state.entries = state.entries.filter (entrada => entrada.id !== id)

}