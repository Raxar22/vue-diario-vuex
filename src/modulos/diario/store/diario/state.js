
export default ()  => ({
    isLoading : true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Voluptate laborum ex sit labore duis excepteur qui adipisicing. Mollit proident qui anim tempor minim sunt ea sint. Elit enim ad et velit mollit eiusmod nostrud aute tempor nostrud consequat nisi.',
            picture: null,
        },
        {
            id: new Date().getTime()+1000,
            date: new Date().toDateString(),
            text: 'Qui Lorem exercitation commodo proident labore consequat mollit nostrud minim. Sit culpa ex voluptate exercitation ea elit pariatur anim in consequat. Ex consectetur non irure deserunt ad incididunt qui nisi. Sint duis cupidatat reprehenderit id ipsum ullamco amet ex.',
            picture: null,
        },
        {
            id: new Date().getTime()+2000,
            date: new Date().toDateString(),
            text: 'Qui cupidatat labore magna consectetur. Mollit dolore laboris ut non culpa enim ipsum tempor. Eiusmod est minim mollit exercitation. Est duis Lorem consectetur enim occaecat duis duis enim deserunt ipsum cupidatat commodo velit deserunt. Culpa aliqua deserunt officia mollit eu proident. Minim aliquip esse ex quis aliquip anim enim anim esse laboris.',
            picture: null,
        },
    ]
})