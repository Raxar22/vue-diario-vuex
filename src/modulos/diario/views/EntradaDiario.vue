<template>
    <template v-if="entradaVista">
        <div  class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{dia}}</span>
                <span class="mx-1 fs-3">{{mes}} </span>
                <span class="mx-2 fs-4 fw-light"> {{anno}}</span>
            </div>

            <div>

                <input  v-show="false" 
                        type="file" 
                        @change="onSelectedImage" 
                        ref="imageSelector" 
                        accept="image/png, image/jpeg, image/jpg">

                <button v-if="entradaVista.id" class="btn btn-danger mx-2"
                @click="borrandoEntrada">
                    Borrar <i class="fa fa-trash-alt"></i>
                </button>   

                <button class="btn btn-primary" @click="onSelectImage">
                    Subir Foto <i class="fa fa-upload"></i>
                </button>   

            </div>

        </div>
        <hr>
        <div class="d-flex flex-column px-3 h-75">
            <textarea v-model="entradaVista.text" placeholder="¿Qué sucedio hoy?"></textarea>
        </div>
    </template>

    <fab 
        icon="fa-save"
        @on:click="guardarEntrada">
    </fab>

    <img    
        v-if="entradaVista.imagen && !localImage" 
        :src="entradaVista.imagen" 
        alt="Imagen entrada" 
        class="img-thubnail">
 

     <img
        v-if="localImage"
        :src="localImage"
        alt="Imagen entrada" class="img-thubnail">



</template>

<script>

import {defineAsyncComponent} from 'vue'
import { mapGetters, mapActions } from 'vuex';
import Swal from 'sweetalert2'

import getDiaMesAnno from '@/modulos/diario/helpers/getDiaMesAnno'
import uploadImage from '@/modulos/diario/helpers/uploadImage'

export default {
    name:'EntradaDiario',
    props:{
        id: {
            type: String,
            required: true
        }
    },
    data(){
        return {
            entradaVista: null,
            localImage: null,
            file: null
        }
    },
    components:{
        Fab: defineAsyncComponent(()=> import('@/modulos/diario/components/Fab.vue'))
    },
    computed:{
      ...mapGetters('diario',['obtenerEntradaById']),
      dia(){ 
        const {dia} = getDiaMesAnno (this.entradaVista.fecha)
        return dia.replace('-','')
      },
      mes(){ 
        const {mes} = getDiaMesAnno (this.entradaVista.fecha)
        return mes
      },
      anno(){ 
        const {anno} = getDiaMesAnno (this.entradaVista.fecha)
        return anno
    },

    },
    methods:{
        cargarEntrada(){
            let Entrada
            if (this.id === 'nuevo')
            {
                Entrada = {
                    text: '',
                    fecha: new Date().getTime()
                }
            }
            else
            {
                Entrada = this.obtenerEntradaById (this.id)
                //console.log (Entrada)
                if (!Entrada) this.$router.push({name: 'NoEntradas'})
            }

            this.entradaVista = Entrada
        },
        ...mapActions('diario',['actualizarEntradas','crearEntradas','borrarEntrada']),
        async guardarEntrada(){
            Swal.fire({
                title: 'Espere por favor',
                allowOutsideClick: false
            })

            Swal.showLoading()


            const picture = await uploadImage (this.file)
            this.entradaVista.imagen = picture

            if (this.entradaVista.id)
            {
                await this.actualizarEntradas(this.entradaVista)
            }
            else
            {
                const id = await this.crearEntradas(this.entradaVista)
                this.$router.push({name: 'Entrada', params: {id: id}})
            }

            this.file =null
            Swal.fire('Guardado','Entrada registrada con éxito','success')

        },
        async borrandoEntrada()
        {
            const resultado = await Swal.fire ({
                title: '¿Está seguro?',
                text: 'Una vez borrado, no se puede recuperar',
                showDenyButton: true,
                confirmButtonText: 'Si, estoy seguro'
            })
            if (resultado.isConfirmed)
            {

                Swal.fire({
                    title: 'Espere por favor',
                    allowOutsideClick: false

                })

                Swal.showLoading()

                await this.borrarEntrada(this.entradaVista.id)
                this.$router.push({name: 'NoEntradas'})

                Swal.fire ('Eliminado','','success')

            }
        },
        onSelectedImage(event)
        {
            const file =event.target.files[0]
            if (!file)
            {
                this.localImage= null
                this.file = null
                return
            } 

            this.file = file
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL (file)
            // console.log(event.target.files[0])
        },
        onSelectImage()
        {
            this.$refs.imageSelector.click()
        }
    },
    created(){
       this.cargarEntrada()
    },
    watch:{
        id(){
            this.cargarEntrada()
        }
    }
}
</script>

<style lang="scss" scoped>

    textarea{
        font-size: 20px;
        border: none;
        height: 100%;
        &:focus {
            outline: none;
        }
    }
    img{
        width: 350px;
        position: fixed;
        bottom: 100px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
    }

</style>