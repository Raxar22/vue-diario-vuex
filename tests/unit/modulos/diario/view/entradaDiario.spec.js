import Swal from 'sweetalert2'
import { shallowMount } from "@vue/test-utils"
import { createStore } from "vuex"

import diario from '@/modulos/diario/store/diario/index'
import { diarioState } from "../../../Mocks/test-diario-state"

import EntradaDiario from '@/modulos/diario/views/EntradaDiario'
import { stringifyQuery } from 'vue-router'


const createVuexStore = (inicialState)=> createStore({
    modules:{
        diario: {
            ...diario,
            state: {...inicialState}
        }
    }
})  

jest.mock('sweetalert2',()=>({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))


describe ('Pruebas en la Entrada del Diario',()=>{


    const store = createVuexStore(diarioState)
    store.dispatch = jest.fn()
    
    const mockRouter = {
        push: jest.fn()
    }


    let wrapper

    beforeEach(()=>{
        jest.clearAllMocks()
        wrapper = shallowMount (EntradaDiario, {
            props:{
                id:'-N8ucyy0iiZQ851RnyNn'
            },
            global:{
                mocks:{
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test ('debe de sacar al usuario porque el id no existe',() => {
        // const wrapper = shallowMount (EntradaDiario, {
        //     props: {
        //         id: 'Easdsdas'
        //     },
        //     global:{
        //         mocks:{
        //             $router: mockRouter
        //         },
        //         plugins: [store]
        //     }
        // })

        // expect (mockRouter.push).toHaveBeenCalledWith({name: 'NoEntradas'})

    })
    test ('debe de mostrar la entrada correctamente',() => {
        expect(wrapper.html()).toMatchSnapshot()
        expect (mockRouter.push).not.toHaveBeenCalled()
    })

    test ('Debe de borrar la entrada y salir', async ()=>{

        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed:true}))

       await wrapper.find ('.btn-danger').trigger('click')

        expect (Swal.fire).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })
        expect (store.dispatch).toHaveBeenCalledWith('diario/borrarEntrada','-N8ucyy0iiZQ851RnyNn')
        expect (mockRouter.push).toHaveBeenCalled()
    })

})