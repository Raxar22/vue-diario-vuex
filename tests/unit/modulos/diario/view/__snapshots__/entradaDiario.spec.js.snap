// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en la Entrada del Diario debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">Lunes  8</span><span class="mx-1 fs-3">Agosto</span><span class="mx-2 fs-4 fw-light">2022</span></div>
  <div><input type="file" accept="image/png, image/jpeg, image/jpg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir Foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedio hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<img src="https://res.cloudinary.com/dbsmm77sk/image/upload/v1659914939/curso_vue/imqz3ektplfjdt47yzvm.jpg" alt="Imagen entrada" class="img-thubnail">
<!--v-if-->
`;
